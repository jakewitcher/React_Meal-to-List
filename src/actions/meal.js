import uuid from 'uuid';

export const addMeal = (
    {
        name = '',
        itemList = [],
    } = {}
) => ({
    type: 'ADD_MEAL',
    meal: {
        id: uuid(),
        name,
        itemList,
    }
});

export const deleteMeal = (id = '') => ({
    type: 'DELETE_MEAL',
    id,
});

export const editMeal = (
    {
        name = '',
        id = '',
        itemList = [],
    } = {}
) => ({
    type: 'EDIT_MEAL',
    meal: {
        id,
        name,
        itemList,
    }
});