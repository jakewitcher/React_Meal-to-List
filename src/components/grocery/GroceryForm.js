import React from 'react';

const GroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        updateGrocery,
        title,
        groceryId,
    } = props;

    const handleGroceryChange = (name, id, items) => {
        updateGrocery({
            name,
            items,
            id,
        });
        resetGrocery();
    };

    return (
        <div className="form">
            <h2 className="form__header">{title}</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => handleGroceryChange(groceryName, groceryId, itemsList)}
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default GroceryForm;