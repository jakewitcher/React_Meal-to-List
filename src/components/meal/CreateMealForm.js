import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        dispatch
    } = props.mealProps;
    const { addMeal } = props;

    return (
        <div className="form">
            <h2 className="form__header">Create a New Meal</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value="pound(s)">pound(s)</option>
                            <option value="ounce(s)">ounce(s)</option>
                            <option value="gallon(s)">gallons</option>
                            <option value="quart(s)">quart(s)</option>
                            <option value="pint(s)">pint(s)</option>
                            <option value="item(s)">item(s)</option>
                            <option value="bag(s)">bag(s)</option>
                            <option value="box(es)">box(es)</option>
                            <option value="container(s)">container(s)</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            dispatch(addMeal({
                                name,
                                itemList
                            }));
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}



export default MealForm;