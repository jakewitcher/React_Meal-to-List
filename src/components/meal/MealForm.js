import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        updateMeal,
        mealId,
        title,
    } = props;

    const pluralUnitValue = (amount, unitType) => {
        if (amount < 2) {
            return `${unitType}`
        }
        return `${unitType}s`
    };

    return (
        <div className="form">
            <h2 className="form__header">{title}</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value={pluralUnitValue(amount, 'pound')}>{pluralUnitValue(amount, 'pound')}</option>
                            <option value={pluralUnitValue(amount, 'ounce')}>{pluralUnitValue(amount, 'ounce')}</option>
                            <option value={pluralUnitValue(amount, 'gallon')}>{pluralUnitValue(amount, 'gallon')}</option>
                            <option value={pluralUnitValue(amount, 'quart')}>{pluralUnitValue(amount, 'quart')}</option>
                            <option value={pluralUnitValue(amount, 'pint')}>{pluralUnitValue(amount, 'pint')}</option>
                            <option value={pluralUnitValue(amount, 'item')}>{pluralUnitValue(amount, 'item')}</option>
                            <option value={pluralUnitValue(amount, 'bag')}>{pluralUnitValue(amount, 'bag')}</option>
                            <option value={pluralUnitValue(amount, 'container')}>{pluralUnitValue(amount, 'container')}</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            const id = mealId;
                            updateMeal({
                                name,
                                itemList,
                                id,
                            });
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}

export default MealForm;