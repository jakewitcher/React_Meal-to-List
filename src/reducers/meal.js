import { defaultMealsList } from './seed';

// const mealDefaultState = {
//     mealList: [],
// };

export default (state = {mealList: defaultMealsList}, action) => {
    switch (action.type) {
        case 'ADD_MEAL':
            return {
                ...state,
                mealList: [...state.mealList, action.meal]
            };
        default:
            return state;
    }
}