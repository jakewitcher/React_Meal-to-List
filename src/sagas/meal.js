import { put, takeEvery, select } from 'redux-saga/effects';
import database from '../firebase/firebase';

const getUserId = (state) => state.auth.uid;

const formatItemList = (list) => {
    return list.reduce((list, item) => {
        list[item.itemName] = {
            amount: item.amount,
            unit: item.unit
        };
        return list;
    }, {});
};

function* addMeal({ meal = {} } = {}) {
    const { itemList, name } = meal;
    const items = formatItemList(itemList);

    let id;
    const uid = yield select(getUserId);
    yield database.ref(`users/${uid}/meals`).push({ name, items }).then((ref) => {
        id = ref.key;
    });

    yield put({
        type: 'ADD_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
}

function* editMeal({ meal = {} }) {
    const { itemList, name, id } = meal;
    const items = formatItemList(itemList);
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).update({ name, items });

    yield put({
        type: 'EDIT_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
};

function* deleteMeal({ id = '' }) {
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).remove();

    yield put({
        type: 'DELETE_MEAL',
        id,
    });
}

function* setMeal() {
    const meals = [];
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals`).once('value').then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
            const itemList = [];
            childSnapshot.child('items').forEach((item) => {
                itemList.push({
                    itemName: item.key,
                    amount: item.val().amount,
                    unit: item.val().unit,
                });
            });

            meals.push({
                id: childSnapshot.key,
                name: childSnapshot.val().name,
                itemList,
            });
        });
    });

    yield put({ type: 'SET_MEAL', meals });
};

export function* watchAddMeal() {
    yield takeEvery('ON_ADD_MEAL', addMeal);
};

export function* watchEditMeal() {
    yield takeEvery('ON_EDIT_MEAL', editMeal);
};

export function* watchDeleteMeal() {
    yield takeEvery('ON_DELETE_MEAL', deleteMeal);
};

export function* watchSetMeal() {
    yield takeEvery('ON_SET_MEAL', setMeal);
}