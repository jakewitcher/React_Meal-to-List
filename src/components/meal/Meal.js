import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { onDeleteMeal } from '../../actions/meal';

export const Meal = (props) => {
    const { name, id, itemList } = props;
    return (
        <div className="list-type">
            <Link className="list-type__link" to={`/meals/meal/${id}`}>
                <p className="list-type__name">{name}</p>
            </Link>
            <p className="list-type__details">number of items: <span className="list-type__details list-type__details--span"> {itemList.length}</span></p>
            <div className="list-type__update">
                <Link className="list-type__link" to={`/meals/edit/${id}`}>
                    <i className="list-type__edit fas fa-edit" />
                </Link>
                <i
                    className="list-type__delete fas fa-trash"
                    onClick={() => props.onDeleteMeal(id)}
                />
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) => {
    return {
        onDeleteMeal: (id) => dispatch(onDeleteMeal(id)),
    }
}

export default connect(undefined, mapDispatchToProps)(Meal);