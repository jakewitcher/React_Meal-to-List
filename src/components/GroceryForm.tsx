import React, { ChangeEvent } from "react";
import { Item } from "../store/item/types";
import { Meal, MealState } from "../store/meal/types";
import { Grocery, GroceryMeals } from "../store/grocery/types";

interface IMealFormProps {
  groceryName: string;
  itemList: [] | Item[];
  groceryId: string;
  selectedMeal: Meal;
  meals: MealState;
  groceryListMeals: GroceryMeals;
  addMealToList(): void;
  selectedMealChange(e: ChangeEvent<HTMLSelectElement>): void;
  groceryNameChange(e: React.FormEvent<HTMLInputElement>): void;
  resetGrocery(): void;
  updateGrocery(grocery: Grocery): void;
}

const GroceryForm: React.FC<IMealFormProps> = props => {
  const {
    selectedMealChange,
    selectedMeal,
    meals,
    groceryName,
    itemList,
    groceryListMeals,
    groceryNameChange,
    addMealToList,
    resetGrocery,
    updateGrocery,
    groceryId
  } = props;

  const handleGroceryChange = (
    name: string,
    itemList: Item[],
    groceryListMeals: GroceryMeals,
    id: string
  ) => {
    updateGrocery({
      name,
      itemList,
      groceryListMeals,
      id
    });
    resetGrocery();
  };

  return (
    <div className="form--grocery">
      <div className="input-box input-box--meal-name">
        <p>Meals</p>
        <select
          className="input input--dropdown"
          name="selectedMeal"
          value={selectedMeal.name}
          onChange={selectedMealChange}
        >
          {meals.mealList.map(meal => (
            <option value={meal.name} key={meal.id}>
              {meal.name}
            </option>
          ))}
        </select>
      </div>
      <div className="button-box__add--meal">
        <button className="button button__add" onClick={addMealToList}>
          Add Meal
        </button>
      </div>
      <div className="input-box__grocery-name">
        <p>Grocery List Name</p>
        <input
          className="input__grocery-name"
          type="text"
          value={groceryName}
          onChange={groceryNameChange}
        />
      </div>
      <div className="button-box__save--grocery">
        <button
          className="button button__save"
          onClick={() =>
            handleGroceryChange(
              groceryName,
              itemList,
              groceryListMeals,
              groceryId
            )
          }
        >
          Save Grocery List
        </button>
      </div>
    </div>
  );
};

export default GroceryForm;
