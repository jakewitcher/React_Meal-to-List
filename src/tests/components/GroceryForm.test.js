import React from 'react';
import { shallow } from 'enzyme';
import GroceryForm from '../../components/grocery/GroceryForm';
import { mealsList } from '../fixtures/meal'

let title, 
    groceryName, 
    groceryNameChange, 
    selectedMeal,
    selectedMealChange,
    addMealToList,
    updateGrocery, 
    resetGrocery,
    meals,
    wrapper;

beforeEach(() => {
    title = 'grocery';
    groceryName = 'Monday Meals';
    selectedMeal = mealsList[0];
    meals = { mealList: mealsList };
    groceryNameChange = jest.fn();
    selectedMealChange = jest.fn();
    addMealToList = jest.fn();
    updateGrocery = jest.fn();
    resetGrocery = jest.fn();
    wrapper = shallow(
        <GroceryForm 
            title ={title}
            meals={meals}
            groceryName ={groceryName}
            groceryNameChange={groceryNameChange}
            selectedMeal={selectedMeal}
            selectedMealChange ={selectedMealChange}
            addMealToList={addMealToList}
            updateGrocery={updateGrocery}
            resetGrocery={resetGrocery}
        />
    );
});

test('should render GroceryForm correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should set grocery name', () => {
    const value = { target: { value: groceryName }};
    wrapper.find('input').at(0).simulate('change', value);
    expect(groceryNameChange).toHaveBeenLastCalledWith(value);
});

test('should set selected meal', () => {
    const value = { target: { value: selectedMeal }};
    wrapper.find('select').simulate('change', value);
    expect(selectedMealChange).toHaveBeenLastCalledWith(value);
});

test('should add selected meal items to grocery list', () => {
    wrapper.find('button').at(0).simulate('click');
    expect(addMealToList).toHaveBeenCalled();
});

test('should add or edit grocery list and reset state', () => {
    wrapper.find('button').at(1).simulate('click');
    expect(updateGrocery).toHaveBeenCalled();
    expect(resetGrocery).toHaveBeenCalled();
});