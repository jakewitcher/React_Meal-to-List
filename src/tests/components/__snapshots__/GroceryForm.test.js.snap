// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render GroceryForm correctly 1`] = `
<div
  className="form"
>
  <h2
    className="form__header"
  >
    grocery
  </h2>
  <div
    className="form__form"
  >
    <div
      className="form__nameinputfield"
    >
      <p>
        Grocery List Name
      </p>
      <input
        className="form__nameinputfield--input"
        onChange={[MockFunction]}
        type="text"
        value="Monday Meals"
      />
    </div>
    <div
      className="form__iteminputfields"
    >
      <div>
        <p>
          Meals
        </p>
        <select
          className="form__iteminputfields--input form__iteminputfields--dropdown"
          name="selectedMeal"
          onChange={[MockFunction]}
          value="breakfast"
        >
          <option
            key="101"
            value="breakfast"
          >
            breakfast
          </option>
          <option
            key="102"
            value="lunch"
          >
            lunch
          </option>
          <option
            key="103"
            value="dinner"
          >
            dinner
          </option>
        </select>
      </div>
      <div
        className="form__button"
      >
        <button
          className="form__button--add"
          onClick={[MockFunction]}
        >
          Add Meal
        </button>
      </div>
    </div>
    <div>
      <button
        className="form__button--save"
        onClick={[Function]}
      >
        Save Grocery List
      </button>
    </div>
  </div>
</div>
`;
